docs/dev/relay_pass_infra.rst

```diff
@@ -0,0 +1,616 @@
+      std::string name;
+      int opt_level;
+      std::vector<std::string> required;
+    };
+
+PassContext
+^^^^^^^^^^^
+
+``PassContext`` carries useful information for an optimization pass. For
+example, it contains the error reporting system so optimization authors can
+provide diagnostics about why an optimization fails. ``PassContext`` is also
+designed to replace the old ``BuildConfig`` which was used to help users
+configure the compilation options, including optimization level and
+required/disabled passes, etc. For instance, we may have a configuration which
+performs all passes at ``opt_level=3`` with some disabled passes using
+``disabled_pass=xx`` provided by ``PassContext``. Now  we could glob all passes
+at ``opt_level=3`` and exclude those in disabled pass list.
+
+This class is designed as a thread local so that users can conveniently use the
+Python ``with`` syntax to perform optimizations under a certain configuration.
+Examples will be provided later to show how we can use both C++ and Python APIs
+to create a compilation pipeline using pass context.
+
+.. code:: c++
+
+    class PassContextNode : public RelayNode {
+     public:
+      ErrorReporter err_reporter;
+      int opt_level{2};
+      int fallback_device{static_cast<int>(kDLCPU)};
+      tvm::Array<tvm::Expr> required_pass;
+      tvm::Array<tvm::Expr> disabled_pass;
+    };
+
+.. code:: c++
+
+    class PassContext : public NodeRef {
+     public:
+      TVM_DLL static PassContext Create();
+      TVM_DLL static PassContext Current();
+      /* Other fields are omitted. */
+    
+     private:
+      // The entry of a pass context scope.
+      TVM_DLL void EnterWithScope();
+      // The exit of a pass context scope.
+      TVM_DLL void ExitWithScope();
+    
+      // Classes to get the Python `with` like syntax.
+      friend class tvm::With<PassContext>;
+    };
+   
+
+Pass Constructs
+^^^^^^^^^^^^^^^
+
+The pass infra is designed in a hierarchical manner, and it could work at
+various granularity of a Relay program. A pure virtual class ``PassNode`` is
+introduced to serve as the base of the different optimization passes. This class
+contains several virtual methods that will be implemented by the
+subclasses at the module level, function level, or a sequence of passes..
+
+.. code:: c++
+
+    class PassNode : RelayNode {
+      virtual PassInfo Info() const = 0;
+      virtual Module operator()(const Module& mod
+                                const PassContext& pass_ctx) const = 0;
+    };
+
+The functor shows how a pass will be realized, i.e. it always works on a `Relay
+module`_ under a certain context. All passes are designed in a ``Module`` to
+``Module`` manner. Therefore, optimizations governed by the pass infra will
+always update the whole module.
+
+Several subclasses are created to implement different types of optimization
+passes, e.g. function-level passes, module-level passes, and sequential passes.
```

Comment:
```suggestion
passes, e.g., function-level passes, module-level passes, and sequential passes.
```
Also it should be explained what "sequential passes" means somewhere. (I presume it means a sequence of passes; is that right?)

Reply:
yes, it is right.